pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+		if object.Spec.VolumePath == volumePath {
+			log.Infof("found CR for VolumePath: %q", volumePath)
+			volumeMigration.volumePathToVolumeInfo.Store(volumePath, object.Spec)
+			log.Debugf("cached VolumeMigrationSpec for VolumePath: %q", volumePath)
+			return object.Spec, nil
+		}
+	}
+	log.Infof("Could not retrieve CnsvSphereVolumeMigrationSpec from cache and K8s CR for Volume Path: %q. volume may not be registered", volumePath)
+	return migration.CnsvSphereVolumeMigrationSpec{}, ErrVolumeNotRegistered
+}
+
+// SaveVolumeInfo helps create CR for given cnsvSphereVolumeMigration
+// this func also update local cache with supplied cnsvSphereVolumeMigration, after successful creation of CR
+func (volumeMigration *volumeMigration) SaveVolumeInfo(ctx context.Context, cnsvSphereVolumeMigration migration.CnsvSphereVolumeMigration) error {
+	log := logger.GetLogger(ctx)
+	log.Infof("creating CR for cnsvSphereVolumeMigration: %+v", cnsvSphereVolumeMigration)
+	err := volumeMigration.migrationClient.Create(ctx, &cnsvSphereVolumeMigration)
+	if err != nil {
+		log.Errorf("failed to create CR for cnsvSphereVolumeMigration. Error: %v", err)
+		return err
+	}
+	log.Infof("successfully created CR for cnsvSphereVolumeMigration: %v", cnsvSphereVolumeMigration)
+	volumeMigration.volumePathToVolumeInfo.Store(cnsvSphereVolumeMigration.Spec.VolumePath, cnsvSphereVolumeMigration.Spec)
+	log.Infof("successfully updated cache for cnsvSphereVolumeMigrationSpec: %v", cnsvSphereVolumeMigration)
+	return nil
+}
+
+// DeleteVolumeInfo helps delete cached mapping of volumePath to VolumeInfo and
+// also helps delete K8s Custom Resource for specified volumePath
+func (volumeMigration *volumeMigration) DeleteVolumeInfo(ctx context.Context, volumePath string) error {
+	log := logger.GetLogger(ctx)
+	volumeMigrationCRList := &migration.CnsvSphereVolumeMigrationList{}
+	err := volumeMigration.migrationClient.List(ctx, volumeMigrationCRList)
+	if err != nil {
+		log.Errorf("failed to get volumeMigration CR list")
+		return err
+	}
+	for _, object := range volumeMigrationCRList.Items {
+		if object.Spec.VolumePath == volumePath {
+			err = volumeMigration.migrationClient.Delete(ctx, &object)
+			if err != nil {
+				log.Errorf("failed to delete CR for volumeMigration having volumePath: %q", volumePath)
+				return err
+			}
+			log.Infof("successfully deleted CR for volumeMigration having volumePath: %q", volumePath)
+			break
+		}
+	}
+	volumeMigration.volumePathToVolumeInfo.Delete(volumePath)
+	log.Infof("successfully deleted volumeInfo from cache for volumePath: %q", volumePath)
+	return nil
+}
+
+// LoadAllVolumeInfo helps cache all CR for CnsvSphereVolumeMigration
+func (volumeMigration *volumeMigration) LoadAllVolumeInfo(ctx context.Context) error {
+	log := logger.GetLogger(ctx)
+	volumeMigrationCRList := &migration.CnsvSphereVolumeMigrationList{}
+	err := volumeMigration.migrationClient.List(ctx, volumeMigrationCRList)
```

Comment:
Listing all migration CRs will be expensive. Why not have a watch on CRD and keep the cache up to date?

Reply:
Yes an informer would definitely help. Also, how are we naming instances? If there's a deterministic way (some function of volumePath maybe?), we could leverage that that do a quicker look up.

Reply:
Let me use the `informer`. `watch` may have to deal with timeouts, disconnects. etc.

Reply:
I am working on this requirement. this is not included in the https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/235/commits/9185f0ba4ae03d3dc4b634e43504b8df83a4898a

Reply:
done. added informer for cache maintenance. removed the logic of listing all CRs.